File: backend/app/schemas/evidence.py
from datetime import datetime, date
from typing import Optional, List, Any
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, Field, model_validator, field_serializer


class EvidenceBase(BaseModel):
    business_id: Optional[str] = Field(None, max_length=64)
    evidence_type: str = Field(...)
    source_type: Optional[str] = Field(default="USER_ENTERED")
    source: Optional[str] = Field(None, max_length=150)
    source_name: Optional[str] = None
    source_url: Optional[str] = None
    provider_id: Optional[str] = None
    provider_record_id: Optional[str] = None
    description: Optional[str] = None
    value: Optional[Decimal] = None
    numeric_value: Optional[Decimal] = None
    text_value: Optional[str] = None
    boolean_value: Optional[bool] = None
    observation_date: Optional[date] = None
    state: Optional[str] = None
    district: Optional[str] = None
    commodity: Optional[str] = None
    market_id: Optional[str] = None
    market_name: Optional[str] = None
    price_type: Optional[str] = None
    currency: Optional[str] = "INR"
    quantity_unit: Optional[str] = None
    provider_original_unit: Optional[str] = None
    content_hash: Optional[str] = None
    confidence: Optional[Decimal] = Field(default=Decimal("1.00"), ge=0, le=1)
    is_observed: Optional[bool] = True
    is_estimated: Optional[bool] = False

    @model_validator(mode="after")
    def validate_types(self):
        if self.evidence_type:
            self.evidence_type = self.evidence_type.upper()
        if self.source_type:
            self.source_type = self.source_type.upper()
        return self


class EvidenceCreate(EvidenceBase):
    id: Optional[str] = Field(None, min_length=1, max_length=64)


class EvidenceResponse(EvidenceBase):
    id: str
    created_at: datetime
    server_revision: Optional[int] = None

    @field_serializer("numeric_value")
    def serialize_numeric_value(self, v: Optional[Decimal]) -> Optional[str]:
        if v is not None:
            return f"{Decimal(str(v)):.4f}"
        return None

    @field_serializer("value")
    def serialize_value(self, v: Optional[Decimal]) -> Optional[str]:
        if v is not None:
            return f"{Decimal(str(v)):.4f}"
        return None

    model_config = ConfigDict(from_attributes=True)


class MarketPriceQueryResponse(BaseModel):
    evidence: Optional[EvidenceResponse] = None
    freshness: str
    from_cache: bool
    status: str
    warnings: List[str] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)

File: model/Evidence.ts
import { Model, Relation } from '@nozbe/watermelondb';
import { field, readonly, date, relation } from '@nozbe/watermelondb/decorators';
import type Business from './Business';

export default class Evidence extends Model {
    /**
     * Represents the PostgreSQL BIGINT monotonic global sync sequence revision.
     * Represented as JS number for current SIH prototype scale without creating a second revision system.
     */
    @field('server_revision') serverRevision!: number;
    static table = 'evidence';
    static associations = {
        businesses: { type: 'belongs_to', key: 'business_id' },
    } as const;

    @field('business_id') businessId!: string;
    
    // Canonical fields
    @field('source_type') sourceType!: string;
    @field('provider_id') providerId!: string;
    @field('provider_record_id') providerRecordId!: string;
    @field('evidence_type') evidenceType!: string;
    @field('source_name') sourceName!: string;
    @field('source_url') sourceUrl!: string;
    @field('metric_name') metricName!: string;
    
    // Values - numeric value stored as string to preserve Decimal precision
    @field('numeric_value') numericValue!: string;
    @field('text_value') textValue!: string;
    @field('boolean_value') booleanValue!: boolean;
    
    // Metadata
    @field('observation_date') observationDate!: number;
    @field('retrieved_at') retrievedAt!: number;
    @field('state') state!: string;
    @field('district') district!: string;
    @field('commodity') commodity!: string;
    @field('market_id') marketId!: string;
    @field('market_name') marketName!: string;
    @field('price_type') priceType!: string;
    @field('currency') currency!: string;
    @field('quantity_unit') quantityUnit!: string;
    @field('content_hash') contentHash!: string;
    @field('derivation_type') derivationType!: string;
    @field('derivation_version') derivationVersion!: string;

    // Legacy fields (kept temporarily for migration compat)
    @field('source') source!: string;
    @field('description') description!: string;
    @field('value') value!: number;
    @field('confidence') confidence!: number;
    @field('is_observed') isObserved!: boolean;
    @field('is_estimated') isEstimated!: boolean;
    
    @readonly @date('created_at') createdAt!: number;

    @relation('businesses', 'business_id') business!: Relation<Business>;
}
